stream.go

```diff
@@ -72,6 +77,18 @@ func (s *stream[T]) WaitNext() T {
 	return s.state.value
 }
 
+func (s *stream[T]) WaitNextCtx(ctx context.Context) (T, error) {
+	select {
+	case <-s.Changes():
+		// always return context.Err() to indicate a cancelled or timed out context
+		return s.Next(), ctx.Err()
```

Comment:
Please note that the context is associated with the call, not with the stream object. What that means is that if you use the same stream with different contexts, or if you use it with and without contexts, there are times it will advance the stream and other times it won't advance the stream.

If you don't like my previous suggestion because there is some duplicate code I suggest the following:

```golang
select {
case <-s.Changes():
    if ctx.Err() == nil {
        return s.Next(), nil
    }
case <-ctx.Done():
}
var zeroVal T
return zeroVal, ctx.Err()
```
This way, the stream advancement will be consistent with the returned error (or lack of).

Reply:
You are right. I didn't think about the usage of the same stream with different contexts etc.